llama_hub/tools/code_interpreter/base.py

```diff
@@ -0,0 +1,81 @@
+"""Code Interpreter tool spec."""
+
+from llama_index.tools.tool_spec.base import BaseToolSpec
+from llama_index.llms.openai import OpenAI
+from llama_index.llms.base import LLM, ChatMessage, MessageRole
+from typing import Optional, List
+import requests
+import urllib.parse
+import sys
+import subprocess
+
+INITIAL_PROMPT = """
+    You are an AI assistant that is used to generate code.
+    When receiving natural language commands you will generate code to complete the required tasks.
+    The code that you generate should always be complete and runnable, as it will be executed immediately
+
+    You will ALWAYS write the function in python. You must import any libraries you use.
+
+    By default, the only response returned by the code is through stdout. Use print statements heavily.
+
+    Any code that is passed in as a prompt is NOT executed.
+
+    If you are requested to generate a graph, and want to use something like plt.show(), DONT. Instead save the output to a file.
+"""
+
+
+class CodeInterpreterToolSpec(BaseToolSpec):
```

Comment:
Isn't this basically an agent?

If we add LLM calling, etc, what's the conceptual difference between a tool and an agent?

Reply:
To me. this is a `PythonCodeAgent`, and a `RouterAgent`. 

the `PythonCodeAgent` has specialized prompts to specialize the agent for code generation, and uses a code interpreter as a tool.